src/Fixer/Casing/MagicMethodCasingFixer.php

```diff
@@ -0,0 +1,226 @@
+        '__get' => '__get',
+        '__invoke' => '__invoke',
+        '__isset' => '__isset',
+        '__set' => '__set',
+        '__set_state' => '__set_state',
+        '__sleep' => '__sleep',
+        '__tostring' => '__toString',
+        '__unset' => '__unset',
+        '__wakeup' => '__wakeup',
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Magic method definitions and calls must be using the correct casing.',
+            [
+                new CodeSample(
+'<?php
+class Foo
+{
+    public function __Sleep()
+    {
+    }
+}
+'
+                ),
+                new CodeSample(
+'<?php
+$foo->__INVOKE(1);
+'
+                ),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isTokenKindFound(T_STRING);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $inClass = 0;
+        $tokenCount = count($tokens);
+
+        for ($index = 1; $index < $tokenCount - 2; ++$index) {
+            if (0 === $inClass && $tokens[$index]->isClassy()) {
+                $inClass = 1;
+                $index = $tokens->getNextTokenOfKind($index, ['{']);
+
+                continue;
+            }
+
+            if (0 !== $inClass) {
+                if ($tokens[$index]->equals('{')) {
+                    ++$inClass;
+
+                    continue;
+                }
+
+                if ($tokens[$index]->equals('}')) {
+                    --$inClass;
+
+                    continue;
+                }
+            }
+
+            if (!$tokens[$index]->isGivenKind(T_STRING)) {
+                continue; // wrong type
+            }
+
+            $content = $tokens[$index]->getContent();
+            if ('__' !== substr($content, 0, 2)) {
+                continue; // cheap look ahead
+            }
+
+            $name = strtolower($content);
+
+            if (!$this->isMagicMethodName($name)) {
+                continue; // method name is not one of the magic ones we can fix
+            }
+
+            $nameInCorrectCasing = $this->getMagicMethodNameInCorrectCasing($name);
+            if ($nameInCorrectCasing === $content) {
+                continue; // method name is already in the correct casing, no fix needed
+            }
+
+            if ($this->isMethodSignature($tokens, $index)) {
+                if (0 !== $inClass) {
+                    $this->setTokenToCorrectCasing($tokens, $index, $nameInCorrectCasing);
+                }
+
+                continue;
+            }
+
+            if ($this->isMethodCall($tokens, $index)) {
+                $this->setTokenToCorrectCasing($tokens, $index, $nameInCorrectCasing);
+
+                continue;
+            }
+
+            if ('__callstatic' === $name && $this->isStaticMethodCall($tokens, $index)) {
```

Comment:
How about `__set_state`?

Reply:
what about it?

Reply:
Is it handled by this fixer? e.g. `Foo::__set_state([]);`

Reply:
it isn't I think... hmmm